server/mqtt.go

```diff
@@ -4958,9 +4947,75 @@ func isMQTTReservedSubscription(subject string) bool {
 	return false
 }
 
+func sparkbReplaceDeathTimestamp(msg []byte) []byte {
+	const VARINT = 0
+	const TIMESTAMP = 1
+
+	orig := msg
+	buf := bytes.NewBuffer(make([]byte, 0, len(msg)+16)) // 16 bytes should be enough if we need to add a timestamp
+	writeDeathTimestamp := func() {
+		// [tck-id-conformance-mqtt-aware-ndeath-timestamp] A Sparkplug Aware
+		// MQTT Server MAY replace the timestamp of NDEATH messages. If it does,
+		// it MUST set the timestamp to the UTC time at which it attempts to
+		// deliver the NDEATH to subscribed clients
+		//
+		// sparkB spec: 6.4.1. Google Protocol Buffer Schema
+		//      optional uint64 timestamp = 1; // Timestamp at message sending time
+		ts := uint64(time.Now().UnixMilli())
+		buf.Write(protoEncodeVarint(TIMESTAMP<<3 | VARINT))
+		buf.Write(protoEncodeVarint(ts))
+	}
+
+	for len(msg) > 0 {
+		numField, typ, size, err := protoScanField(msg)
+		if err != nil {
+			return orig
+		}
+		if typ != VARINT || numField != TIMESTAMP {
+			// Add the field as is
+			buf.Write(msg[:size])
+			msg = msg[size:]
+			continue
+		}
+
+		writeDeathTimestamp()
+
+		// Add the rest of the message as is, we are done
+		buf.Write(msg[size:])
+		return buf.Bytes()
+	}
+
+	// Add timestamp if we did not find one.
+	writeDeathTimestamp()
+
+	return buf.Bytes()
+}
+
 // Common function to mqtt delivery callbacks to serialize and send the message
 // to the `cc` client.
-func (c *client) mqttEnqueuePublishMsgTo(cc *client, sub *subscription, pi uint16, qos byte, dup, retain bool, topic, msg []byte) {
+func (c *client) mqttEnqueuePublishMsgTo(cc *client, sub *subscription, pi uint16, qos byte, dup bool, topic, msg []byte) {
+	// [tck-id-conformance-mqtt-aware-nbirth-mqtt-retain] A Sparkplug Aware
+	// MQTT Server MUST make NBIRTH messages available on the topic:
+	// $sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id with
+	// the MQTT retain flag set to true
+	//
+	// [tck-id-conformance-mqtt-aware-dbirth-mqtt-retain] A Sparkplug Aware
+	// MQTT Server MUST make DBIRTH messages available on the topic:
+	// $sparkplug/certificates/namespace/group_id/DBIRTH/edge_node_id/device_id
+	// with the MQTT retain flag set to true
+	//
+	// $sparkplug/certificates messages are sent as NATS messages, so we
+	// need to add the retain flag when sending them to MQTT ciients.
+
+	retain := false
+	isBirth, isDeath, isCertificate := sparkbParseBirthDeathTopic(topic)
+	if isBirth && isCertificate {
+		retain = true
+	}
+	if isDeath && !isCertificate {
```

Comment:
Can a message be a birth and death? In other words, should we have a `else` instead of separate `if`?

Reply:
Is the concern here the performance or intention revealing? There are 3 "interesting" states here - `do-not-change`, `DEATH` (not certificate), or `BIRTH certificate`. Since DEATH messages are not supposed to appear in the certificates subject space, 0/5 I could express this as:
```go
switch {
case isBirth:
    retain = isCertificate
case isDeath && !isCertificate:
    msg = sparkbReplaceDeathTimestamp(msg)
default:
    // no change
}
```

Reply:
It was just to say that if a message cannot be both `isBirth` and `isDeath`, then logically speaking it is one or the other, so more like `if .. else ..`. That being said, it is true that in Golang, there is extensive use of `switch` instead of `if`, but I guess it is a matter of clarity/conciseness. 
```
if isBirth && isCertificate {
    retain = true
} else if isDeath && !isCertificate {
    msg = sparkbReplaceDeathTimestamp(msg)
}
```

Reply:
k, then really,
```go
if isBirth {
   retain = isCertificate
} else ...same
```

will change.